### PR TITLE
Accelerate Jordan-Wigner transform

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
Jordan-Wigner transforms can take a while to compute. Most of the time is spent in `__init__` of the various Pauli gates.

**Description of the Change:**
Cache Pauli gate objects.

**Benefits:**
Faster transforms (e.g. 160->38 sec. for generating the molecular Hamiltonian of an H2 molecule in the "cc-pvdz" basis). 

**Possible Drawbacks:**
Deepcopy necessary?

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/qchem/observable_hf.py [hunk 1]
```diff
@@ -117,9 +118,16 @@ def qubit_observable(o_ferm, cutoff=1.0e-12):
                 ops = ops + op[1]
                 coeffs = qml.math.concatenate([coeffs, qml.math.array(op[0]) * o_ferm[0][n]])
 
-    o_qubit = simplify(qml.Hamiltonian(coeffs, ops), cutoff=cutoff)
+    return simplify(qml.Hamiltonian(coeffs, ops), cutoff=cutoff)
 
-    return o_qubit
+
+@lru_cache
+def get_pauli_map(n):
```
Comment (soranjh) [inline pennylane/qchem/observable_hf.py:125]: This should be a private function?

**Review by soranjh — COMMENTED** [review]
Thanks @vincentmr for the improvement! Left a suggestion and also a timing comparison for reference (the molecules are `H2, H3+, LiH, H2O`). 

image

File: pennylane/qchem/observable_hf.py (hunk 1 above)
Comment (vincentmr) [reply pennylane/qchem/observable_hf.py:125]: You mean naming it `_get_pauli_map`? Yes probably. I think we could even move it to `pennylane.pauli.utils`, no?

**Comment by vincentmr** [comment]
> Thanks @vincentmr for the improvement! Left a suggestion and also a timing comparison for computing molecular Hamiltonians (the molecules are `H2, H3+, LiH, H2O`).

This is nice, thanks for the timing data.

File: pennylane/qchem/observable_hf.py (hunk 1 above)
Comment (soranjh) [reply pennylane/qchem/observable_hf.py:125]: Yes, it can be renamed to `_get_pauli_map` and moved to `pennylane.pauli.utils`.

**Comment by vincentmr** [comment]
For the record, after the current changes, the biggest bottlenecks are `_pauli_mult` called in `jordan_wigner` and `simplify` called in `qubit_observable`.

**Review by soranjh — APPROVED** [review]

File: pennylane/pauli/utils.py [hunk 2]
```diff
@@ -1379,24 +1379,23 @@ def _pauli_mult(p1, p2):
...
     c = 1.0
 
-    t1 = [t[0] for t in p1]
-    t2 = [t[0] for t in p2]
-
+    t1 = np.array([t[0] for t in p1], dtype=int).reshape((1, -1))
+    t2 = np.array([t[0] for t in p2], dtype=int).reshape((-1, 1))
+    d = (t1 - t2) != 0
+    t1notint2 = np.all(d, axis=0)
+    t2notint1 = np.all(d, axis=1)
```
Comment (soranjh) [inline pennylane/pauli/utils.py:1386]: Seems that these new changes slightly decrease the performance?

Comment (vincentmr) [reply pennylane/pauli/utils.py:1386]: Hi Soran, could you please send me the script you use to benchmark? I would like to test it on my end (+ for future use), and then we can compare results.

Comment (vincentmr) [reply pennylane/pauli/utils.py]: You are right. I think some processes running in parallel on my laptop messed up my numbers. I'll just revert and let's merge the changes as is for now. As you said, `_pauli_mult` can be dealt with later.

**Review by obliviateandsurrender — COMMENTED** [review]
Thanks @vincentmr, overall looks good! Does caching have any ill-effect on memory profiles for building larger observables? It might be nice to have some memory-benchmarks too.

**Comment by vincentmr** [comment]
> Thanks @vincentmr, overall looks good! Does caching have any ill-effect on memory profiles for building larger observables? It might be nice to have some memory-benchmarks too.

Calling `_get_pauli_map` for 4 up to 32 wires, the time for the first call grows linearly, but then it is pretty much constant.
```
mprof: Sampling memory every 0.1s
running new process
time(4)=0.0002491474151611328
time(4)=4.76837158203125e-07
time(8)=0.00027823448181152344
time(8)=7.152557373046875e-07
time(12)=0.0004432201385498047
time(12)=4.76837158203125e-07
time(16)=0.0005674362182617188
time(16)=4.76837158203125e-07
time(20)=0.0006365776062011719
time(20)=4.76837158203125e-07
time(24)=0.0007550716400146484
time(24)=4.76837158203125e-07
time(28)=0.0008456707000732422
time(28)=4.76837158203125e-07
time(32)=0.0014255046844482422
time(32)=9.5367431640625e-07
```
There is no noticeable differences between the memory profiles for any wire count. They all look like (on my laptop)
Figure_1

File: pennylane/pauli/utils.py [hunk 3]
```diff
@@ -1482,3 +1480,15 @@ def _binary_matrix(terms, num_qubits, wire_map=None):
                 binary_matrix[idx][wire_map[wire]] = 1
 
     return binary_matrix
+
+
+@lru_cache
+def _get_pauli_map(n):
+    r"""Return Pauli gate objects on wires 0 up to n.
```
Comment (obliviateandsurrender) [inline pennylane/pauli/utils.py:1487]:
```suggestion
    r"""Return a list of Pauli operator objects acting on wires `0` up to `n`.
```

File: pennylane/qchem/observable_hf.py [hunk 4]
```diff
@@ -193,14 +189,15 @@ def jordan_wigner(op, notation="physicist"):
...
         if len(term) == 0:
             o[i] = qml.Identity(0)
         if len(term) == 1:
-            o[i] = pauli_map[term[0][1]](term[0][0])
+            o[i] = pauli_map[term[0][0]][term[0][1]]
         if len(term) > 1:
-            k = [pauli_map[t[1]](t[0]) for t in term]
-            o[i] = reduce(lambda x, y: x @ y, k)
+            k = [pauli_map[t[0]][t[1]] for t in term]
+            o[i] = Tensor(*k)
```
Comment (obliviateandsurrender) [inline pennylane/qchem/observable_hf.py:201]: This should also work for `len(term) == 1` case, right? If so, we might be able to get rid of one of the upper condition.

**Review by obliviateandsurrender — CHANGES_REQUESTED** [review]

**Review by obliviateandsurrender — APPROVED** [review]
:shipit:

**Pull request merged**